storage/repair.go

```diff
@@ -0,0 +1,283 @@
+	)
+
+	// Record total number of series and total number of blocks
+	totalScope.Counter("series").Inc(diffRes.NumSeries)
+	totalScope.Counter("blocks").Inc(diffRes.NumBlocks)
+	r.logger.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("shard", shard.ID()),
+		xlog.NewLogField("numSeries", diffRes.NumSeries),
+		xlog.NewLogField("numBlocks", diffRes.NumBlocks),
+	).Infof("repair total count")
+
+	// Record size differences
+	sizeDiffSeries := diffRes.SizeDifferences.Series()
+	numBlocks := 0
+	for _, series := range sizeDiffSeries {
+		numBlocks += len(series.Blocks())
+	}
+	sizeDiffScope.Counter("series").Inc(int64(len(sizeDiffSeries)))
+	sizeDiffScope.Counter("blocks").Inc(int64(numBlocks))
+	r.logger.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("shard", shard.ID()),
+		xlog.NewLogField("numSeries", len(sizeDiffSeries)),
+		xlog.NewLogField("numBlocks", numBlocks),
+	).Infof("repair size difference count")
+
+	// Record checksum differences
+	checksumDiffSeries := diffRes.ChecksumDifferences.Series()
+	numBlocks = 0
+	for _, series := range checksumDiffSeries {
+		numBlocks += len(series.Blocks())
+	}
+	checksumDiffScope.Counter("series").Inc(int64(len(checksumDiffSeries)))
+	checksumDiffScope.Counter("blocks").Inc(int64(numBlocks))
+	r.logger.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("shard", shard.ID()),
+		xlog.NewLogField("numSeries", len(checksumDiffSeries)),
+		xlog.NewLogField("numBlocks", numBlocks),
+	).Infof("repair checksum difference count")
+}
+
+type repairFn func() error
+
+type sleepFn func(d time.Duration)
+
+type dbRepairer struct {
+	sync.Mutex
+
+	database      database
+	opts          Options
+	ropts         repair.Options
+	shardRepairer databaseShardRepairer
+
+	repairFn            repairFn
+	sleepFn             sleepFn
+	nowFn               clock.NowFn
+	logger              xlog.Logger
+	repairInterval      time.Duration
+	repairTimeOffset    time.Duration
+	repairCheckInterval time.Duration
+	closed              bool
+}
+
+func newDatabaseRepairer(database database) (databaseRepairer, error) {
+	opts := database.Options()
+	ropts := opts.RepairOptions()
+	if ropts == nil {
+		return nil, errNoRepairOptions
+	}
+	if err := ropts.Validate(); err != nil {
+		return nil, err
+	}
+
+	r := &dbRepairer{
+		database:            database,
+		opts:                opts,
+		ropts:               ropts,
+		shardRepairer:       newShardRepairer(opts, ropts),
+		sleepFn:             time.Sleep,
+		nowFn:               opts.ClockOptions().NowFn(),
+		logger:              opts.InstrumentOptions().Logger(),
+		repairInterval:      ropts.RepairInterval(),
+		repairTimeOffset:    ropts.RepairTimeOffset(),
+		repairCheckInterval: ropts.RepairCheckInterval(),
+	}
+	r.repairFn = r.Repair
+
+	return r, nil
+}
+
+func (r *dbRepairer) Start() {
+	if r.repairInterval <= 0 {
+		return
+	}
+
+	var curIntervalStart time.Time
+
+	for {
+		r.Lock()
```

Comment:
I started with `RWMutex` but the only time we need a write lock is when the database is closed (which only happens during server shutdown), so a `Mutex` should be more performant in this case.